Added 'How the Outbox Works' section and expand requirements with int… (#4875)
dapr/docs: daprdocs/content/en/developing-applications/building-blocks/state-management/howto-outbox.md

Reviewer: Add a "Next Steps" section into this article and put a link to this blog post https://www.diagrid.io/blog/how-dapr-outbox-eliminates-dual-writes-in-distributed-applications.

With Dapr's outbox support, you can notify subscribers when an application's state is created or updated when calling Dapr's [transactions API]({{% ref "state_api.md#state-transactions" %}}).

The diagram below is an overview of how the outbox feature works at a high level:

1) Service A saves/updates state to the state store using a transaction.
2) A message is written to the broker under the same transaction. When the message is successfully delivered to the message broker, the transaction completes, ensuring the state and message are transacted together.
3) The message broker delivers the message topic to any subscribers - in this case, Service B.

<img src="/images/state-management-outbox.png" width=800 alt="Diagram showing the overview of outbox pattern">

> ## How the Outbox Works Under the Hood

Reviewer: ```suggestion
## How outbox works under the hood
```

> Dapr Outbox processes requests in two flows: the user request flow and the background message flow. Together, they guarantee that state and events stay consistent.

Reviewer: ```suggestion
Dapr outbox processes requests in two flows: the user request flow and the background message flow. Together, they guarantee that state and events stay consistent.
```

<img src="/images/state-management-outbox-steps.png" width=800 alt="Diagram showing the steps of the outbox pattern">

> Here is the full sequence of interactions:

Reviewer: ```suggestion
This is the sequence of interactions:
```

> 1. Application calls the Dapr State API to write state transactionally.  

Reviewer: ```suggestion
1. An application calls the Dapr State Management API to write state transactionally using the transactional methods.  
```

   This is the entry point where business data, such as an order or profile update, is submitted for persistence.

2. Dapr publishes an intent message with a unique transaction ID to an internal outbox topic.  
   This durable record ensures the event intent exists before any database commit happens.

> 3. State and a transaction marker are written atomically in the same state store.  

Reviewer: ```suggestion
3. The state and a transaction marker are written atomically in the same state store.  
```

   Both the business data and the marker are committed in the same transaction, preventing partial writes.

> 4. Application receives a success response after the commit.  

Reviewer: ```suggestion
4. The application receives a success response after the transaction commits.  
```

   At this point, the application can continue, knowing state is saved and the event intent is guaranteed.

> 5. Background subscriber reads the intent message.  

Reviewer: ```suggestion
5. A B=background subscriber reads the intent message.  
```

   When outbox is enabled, Dapr starts consumers that process the internal outbox topic.

> 6. Subscriber verifies the transaction marker in the state store.  

Reviewer: ```suggestion
6. The subscriber verifies the transaction marker in the state store.  
```

   This check confirms that the database commit was successful before publishing externally.

7. Verified business event is published to the external pub/sub topic.  
   The event is sent to the configured broker (Kafka, RabbitMQ, etc.) where other services can consume it.

> 8. Marker is cleaned up (deleted) from the state store.  

Reviewer: ```suggestion
8. The marker is cleaned up (deleted) from the state store.  
```
